### ISSUE TITLE
Mixed precision with existing Equinox modules

### ISSUE THREAD

**Issue by samuelstevens** [patrick-kidger/equinox#553]
Hi, I am trying to use [jmp](https://github.com/google-deepmind/jmp) to implement mixed precision in a transformer model in Equinox. These [pytorch docs](https://pytorch.org/docs/stable/amp.html#cuda-op-specific-behavior) have a list of fp16-safe ops and fp16-unsafe ops.

Since softmax is not safe in fp16, I'd like to make sure that all softmaxes are in fp32. If I'm using [eqx.nn.MultiheadAttention](https://docs.kidger.site/equinox/api/nn/attention/#equinox.nn.MultiheadAttention), how can I make sure the softmax runs in fp32? 

```py
# x is a sequence of token vectors in fp32
# self.mhsa is an instance of eqx.nn.MultiheadAttention

x_ln = jax.vmap(self.layernorm1)(x)

# cast to fp16
x_ln, mhsa = self.policy.cast_to_compute((x_ln, self.mhsa))

# how do I make sure the softmax is run in fp32?
post_attn = self.mhsa(query=x_ln, key_=x_ln, value=x_ln, key=attn_key)  

# this will recast to fp32 because x is still fp32
x = x + post_attn  
```

**Comment by patrick-kidger** [comment]
The simplest approach would be to copy-paste the `eqx.nn.MultiheadAttention` implementation, and explicitly add in your desired casts before the softmax (`some_array = some_array.astype(jnp.float32)`).

(Calling `.astype` is actually all `policy.cast_to_compute` is already doing for `x_ln` and the parameters of `mhsa`: see the [jmp implementation here](https://github.com/google-deepmind/jmp/blob/5dfab8e2e3f086740bed961f8a18402e5eb4da12/jmp/_src/policy.py#L27-L61).)

That said, this is a common enough scenario that I think it'd make sense to add a `softmax_dtype` argument to the built-in `eqx.nn.MultiheadAttention` implementation, and probably also a warning if the softmax is happening with too low of a precision to be safe. I would be happy to take a PR on that.
